### PR TITLE
feat: self deleting messages

### PR DESCRIPTION
**Pull request by trOnk12**
---------

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Epic PR consisting of smaller PRs:
- #1552 
- #1592
- #1599
- #1585

This PR exposes the functionality to the client to enqueue messages for self-deletion. The self-deletion is also restarted whenever client loses the reference to UserSessionScope and recreates it.

### Dependencies (Optional)

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by gongracr — COMMENTED** [review]
Looking great after a quick review! 👏  
Just some idea on how to initially tackle a message deletion failure edge case and a comment on some missing tests.

**Comment by vitorhugods** [comment]
Converting to draft as it is not ready for review yet and more PRs are incoming so they can be bundled within this, for example:

- #1584

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EnqueueMessageSelfDeletionUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.kalium.logic.data.id.ConversationId
+
+/**
+ * This use case enqueue the self deletion for a specific conversation id and message id
```
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EnqueueMessageSelfDeletionUseCase.kt]: added, however Message cannot be resolved in a comment 🤔

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                    }
+
+                    delay(timeLeftForDeletion())
+                }
+
+                onGoingSelfDeletionMessagesMutex.withLock {
+                    onGoingSelfDeletionMessages - message.conversationId to message.id
+                }
+
+                messageRepository.deleteMessage(message.id, message.conversationId)
```
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt]: I am not sure how to proceed with this one as for now @gongracr @vitorhugods , if we fail on deleting a message what could we do ? This is a last resort , like if we cannot delete a message there is not much of option then to try what Gonzo is suggesting Should I add Gonzo idea to the PR? Or are we good for now with merging this one as it is ?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EnqueueMessageSelfDeletionUseCase.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EnqueueMessageSelfDeletionUseCase.kt]: Just add an `import` at the top of the file, or type the full canonical name (package + class), like `com.wire.kalium.logic.data.message.Message`, or whatever it is.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/EnqueueMessageSelfDeletionUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.kalium.logic.data.id.ConversationId
+import kotlinx.coroutines.test.TestScope
+import kotlinx.coroutines.test.runTest
+import kotlin.test.Test
+
+class EnqueueMessageSelfDeletionUseCase {
```
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/EnqueueMessageSelfDeletionUseCase.kt]: deleted as it does little to nothing

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+internal class EphemeralMessageDeletionHandlerImpl(
+    private val messageRepository: MessageRepository,
+    private val kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+    userSessionCoroutineScope: CoroutineScope
+) : EphemeralMessageDeletionHandler, CoroutineScope by userSessionCoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = kaliumDispatcher.default
+
+    private val onGoingSelfDeletionMessagesMutex = Mutex()
+    private val onGoingSelfDeletionMessages = mutableMapOf<Pair<ConversationId, String>, Unit>()
```
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt]: I do not think this is going to work since it will have a key per conversation, but we want to have it grouped per messages, so conversation + message id is kind of a merged key making every message unique

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt]: It is actually a deletion so the chances of anything going wrong are even smaller or even not exsiting

[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt]: Full deletion? Don't they stay visible but with the content removed like in the current Scala app? 🤔 

I guess it makes sense tho. When being extra-cautious and having self-deleting messages, there's no point in keeping history of who/when messages were sent. Just erase _everything_.

[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt]: Full deletion, on the receiver side, same is happening on MacOS

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/conversation/message/ApplicationMessageHandler.kt [hunk 5]
```diff
@@ -204,42 +212,33 @@ internal class ApplicationMessageHandlerImpl(
...
-            is MessageContent.RestrictedAsset -> TODO()
         }
     }
 
-    private suspend fun handleKnock(message: Message.Regular) {
-        persistMessage(message)
-    }
-
-    private suspend fun handleTextMessage(
+    private suspend fun adjustTextMessageWithQoutedReference(
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/conversation/message/ApplicationMessageHandler.kt:228]:
```suggestion
    private suspend fun adjustTextMessageWithQuotedReference(
```

As seen in [here](https://github.com/wireapp/kalium/pull/1552/files/e89bcf5c5cb6ed668eb3aae40254d0243059eac1#r1132345115).

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt]: I am going to resolve this as for now and see how it is going, I think in majority we agree that is pretty safe also the last-resort is already implemented when enqueue pending messages to deletion.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt [hunk 6]
```diff
@@ -376,6 +373,42 @@ class ProtoContentMapperImpl(
...
+                    Asset(
+                        original = ephemeralContent.value.original,
+                        status = ephemeralContent.value.status,
+                        expectsReadConfirmation = ephemeralContent.value.expectsReadConfirmation
+                    )
+                )
+                unpackAsset(genericAssetContent)
+            }
+
+            // TODO: Mateusz : for now we ignore anything else then Text
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt:403]: Not up-to-date anymore as `Asset` is handled.

AFAIK, the only possible self-deleting messages are:
- Text
- Asset
- Location

We do not support `Location` yet. So we can just remove this TODO. Or make it like:

```suggestion
            // TODO: Handle self-deleting Location messages when they are implemented
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/ephemeral/EphemeralMessageDeletionHandler.kt [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import kotlinx.datetime.Clock
+import kotlin.coroutines.CoroutineContext
+
+interface EphemeralMessageDeletionHandler {
+
+    fun startSelfDeletion(conversationId: ConversationId, messageId: String)
+    fun enqueuePendingSelfDeletionMessages()
+}
+
+// TODO:Mateusz: implement failure logic (needs to be discussed with a team)
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/ephemeral/EphemeralMessageDeletionHandler.kt:24]: No need to sign `TODOs` btw. We can always check git blame 😄.

No need to change now that it's done, tho.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/conversation/message/ApplicationMessageHandler.kt (hunk 5 above)
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/conversation/message/ApplicationMessageHandler.kt:228]: To be honest. I preferred the old name.

From the caller (`processMessage`) point of view, it is just calling `handleTextMessage`. It doesn't need to know WHAT in a text message should be processed.

By changing this name, now `processMessage` needs to understand how the text message should be processed, instead of just delegating to specialised function.

If clarity was the reason for changing this name, I'd prefer a simple comment inside the function, or a nested function like so:

```kotlin
    private suspend fun handleTextMessage(
        message: Message.Regular,
        messageContent: MessageContent.Text
    ): Message.Regular {
        fun adjustedQuoteReference(): QuoteReference {
            val quotedReference = messageContent.quotedMessageReference
            return if (quotedReference != null) {
                verifyMessageQuote(quotedReference, message)
            } else {
                messageContent.quotedMessageReference
            }
        }
        return message.copy(
            content = messageContent.copy(quotedMessageReference = adjustedQuoteReference())
        )
    }
```

**Review by vitorhugods — COMMENTED** [review]
Looks pretty good to me. Just one thing that slipped away from the previous PRs.

**Review by gongracr — APPROVED** [review]
Great job and thanks for the changes 🙏🌼 However I think it's still worth it to give it a future look to the internal structure we are using to store the enqueued messages to be self-deleted. But since the current priority is to push this forward I'll unblock it and approve 🔥

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/ephemeral/EphemeralMessageDeletionHandler.kt (hunk 7 above)
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/ephemeral/EphemeralMessageDeletionHandler.kt:24]: done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt (hunk 6 above)
Comment (trOnk12) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/ProtoContentMapper.kt:403]: done

**Review by vitorhugods — APPROVED** [review]
😍😍😍😍😍
